test/addr_manager_suite.go

```diff
@@ -0,0 +1,280 @@
+package test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+
+	pstore "github.com/libp2p/go-libp2p-peerstore"
+)
+
+var addressBookSuite = map[string]func(book pstore.AddrBook) func(*testing.T){
+	"Addresses":            testAddresses,
+	"Clear":                testClearWorks,
+	"SetNegativeTTLClears": testSetNegativeTTLClears,
+	"UpdateTTLs":           testUpdateTTLs,
+	"NilAddrsDontBreak":    testNilAddrsDontBreak,
+	"AddressesExpire":      testAddressesExpire,
+}
+
+type AddrMgrFactory func() (pstore.AddrBook, func())
+
+func TestAddrMgr(t *testing.T, factory AddrMgrFactory) {
+	for name, test := range addressBookSuite {
+		// Create a new peerstore.
+		ab, closeFunc := factory()
+
+		// Run the test.
+		t.Run(name, test(ab))
+
+		// Cleanup.
+		if closeFunc != nil {
+			closeFunc()
+		}
+	}
+}
+
+func testAddresses(m pstore.AddrBook) func(*testing.T) {
```

Comment:
This one test is actually many tests and for three separate functions, AddAddr, AddAddrs, ClearAddr.
I would like to see something that gives a lot more granularity to what is actually failing, such as the following:

```go
var ids := []peer.ID{...}
var addrs := []ma.Multiaddr{...}

func testAddAddr(ab pstore.AddrBook) func(*testing.T) {
    return func(t *testing.T) {
	testcases := []struct {
		ab    pstore.AddrBook
		p     peer.ID
		addrs []ma.Multiaddr
		ttl   time.Duration
		want  []ma.Multiaddr
	}{
		{ab, ids[0], []ma.Multiaddr{addrs[0]}, time.Hour, []ma.Multiaddr{addrs[0]}},
		{ab, ids[1], []ma.Multiaddr{addrs[2]}, time.Hour, []ma.Multiaddr{addrs[2]}},
		{
			ab,
			ids[1],
			[]ma.Multiaddr{addrs[2], addrs[3], addrs[4], addrs[4]},
			time.Hour,
			[]ma.Multiaddr{addrs[2], addrs[3], addrs[4]},
		},
	}
	for _, tt := range testcases {
		t.Run("add single address", func(t *testing.T) {
			tt.ab.AddAddr(tt.p, tt.addrs[0], tt.ttl)

			testHas(t, tt.want, tt.ab.Addrs(tt.p))
		})
		t.Run("idempotent address add", func(t *testing.T) {
			tt.ab.AddAddr(tt.p, tt.addrs[0], tt.ttl)
			tt.ab.AddAddr(tt.p, tt.addrs[0], tt.ttl)

			testHas(t, tt.want, tt.ab.Addrs(tt.p))
		})
		t.Run("idempotent multi address add", func(t *testing.T) {
			if len(tt.addrs) < 2 {
				return
			}
			for _, a := range tt.addrs {
				tt.ab.AddAddr(tt.p, a, tt.ttl)
			}
			// add the last addrs twice
			tt.ab.AddAddr(tt.p, tt.addrs[len(tt.addrs)-1], tt.ttl)

			testHas(t, tt.want, tt.ab.Addrs(tt.p))
		})
	}
}
```

Reply:
@lanzafame – I've only moved the tests around in this PR, I haven't introduced any new tests nor modified existing ones.

How about we make this change in the [next PR up for review](https://github.com/libp2p/go-libp2p-peerstore/pull/32)? It precisely touches tests to increase coverage ;-)

Reply:
Done: https://github.com/libp2p/go-libp2p-peerstore/pull/32/commits/027d289a66c6b11d0ac42acabff561d9557b7886